File: tests/human/x/fs/group_by_having.fs
open System
open System.Text.Json

type Person = { name: string; city: string }

let people = [
    { name = "Alice"; city = "Paris" }
    { name = "Bob"; city = "Hanoi" }
    { name = "Charlie"; city = "Paris" }
    { name = "Diana"; city = "Hanoi" }
    { name = "Eve"; city = "Paris" }
    { name = "Frank"; city = "Hanoi" }
    { name = "George"; city = "Paris" }
]

let big =
    people
    |> List.groupBy (fun p -> p.city)
    |> List.filter (fun (_, grp) -> List.length grp >= 4)
    |> List.map (fun (city, grp) -> {| city = city; num = List.length grp |})

let jsonStr = JsonSerializer.Serialize(big)
printfn "%s" jsonStr

File: tests/human/x/clj/group_by_having.clj
(ns group-by-having
  (:require [clojure.data.json :as json]))

(def people
  [{:name "Alice" :city "Paris"}
   {:name "Bob" :city "Hanoi"}
   {:name "Charlie" :city "Paris"}
   {:name "Diana" :city "Hanoi"}
   {:name "Eve" :city "Paris"}
   {:name "Frank" :city "Hanoi"}
   {:name "George" :city "Paris"}])

(def big
  (->> people
       (group-by :city)
       (map (fn [[city grp]] {:city city :num (count grp)}))
       (filter #(>= (:num %) 4))))

(println (json/write-str big))
